Signal/src/ViewControllers/ConversationView/ConversationViewModel.m

```diff
@@ -0,0 +1,1284 @@
+        OWSFailDebug(@"hasMalformedRowChange");
+        // resetMappings will call delegate.conversationViewModelDidUpdate.
+        [self resetMappings];
+        return;
+    }
+
+    NSUInteger oldViewItemCount = self.viewItems.count;
+    if (![self reloadViewItems]) {
+        // These errors are rare.
+        OWSFailDebug(@"could not reload view items; hard resetting message mappings.");
+        // resetMappings will call delegate.conversationViewModelDidUpdate.
+        [self resetMappings];
+        return;
+    }
+
+    OWSLogVerbose(@"self.viewItems.count: %zd -> %zd", oldViewItemCount, self.viewItems.count);
+
+    [self updateViewWitholdItemIdList:oldItemIdList
+                       updatedItemSet:updatedItemSet
+                     oldViewItemCount:oldViewItemCount
+               updatedNeighborItemSet:updatedNeighborItemSet];
+}
+
+- (void)updateViewWitholdItemIdList:(NSArray<NSString *> *)oldItemIdList
+                     updatedItemSet:(NSSet<NSString *> *)updatedItemSet
+                   oldViewItemCount:(NSUInteger)oldViewItemCount
+             updatedNeighborItemSet:(nullable NSMutableSet<NSString *> *)updatedNeighborItemSet
+{
+    OWSAssertDebug(oldItemIdList);
+    OWSAssertDebug(updatedItemSet);
+
+    if (oldItemIdList.count != [NSSet setWithArray:oldItemIdList].count) {
+        OWSFailDebug(@"Old view item list has duplicates.");
+        [self.delegate conversationViewModelDidUpdate:ConversationUpdate.reloadUpdate];
+        return;
+    }
+
+    NSMutableArray<NSString *> *newItemIdList = [NSMutableArray new];
+    NSMutableDictionary<NSString *, id<ConversationViewItem>> *newViewItemMap = [NSMutableDictionary new];
+    for (id<ConversationViewItem> viewItem in self.viewItems) {
+        [newItemIdList addObject:viewItem.itemId];
+        newViewItemMap[viewItem.itemId] = viewItem;
+    }
+
+    if (newItemIdList.count != [NSSet setWithArray:newItemIdList].count) {
+        OWSFailDebug(@"New view item list has duplicates.");
+        [self.delegate conversationViewModelDidUpdate:ConversationUpdate.reloadUpdate];
+        return;
+    }
+
+    NSSet<NSString *> *oldItemIdSet = [NSSet setWithArray:oldItemIdList];
+    NSSet<NSString *> *newItemIdSet = [NSSet setWithArray:newItemIdList];
+
+    // We use sets and dictionaries here to ensure perf.
+    NSMutableSet<NSString *> *deletedItemIdSet = [oldItemIdSet mutableCopy];
+    [deletedItemIdSet minusSet:newItemIdSet];
+    NSMutableSet<NSString *> *insertedItemIdSet = [newItemIdSet mutableCopy];
+    [insertedItemIdSet minusSet:oldItemIdSet];
+    NSArray<NSString *> *deletedItemIdList = [deletedItemIdSet.allObjects sortedArrayUsingSelector:@selector(compare:)];
+    NSArray<NSString *> *insertedItemIdList =
+        [insertedItemIdSet.allObjects sortedArrayUsingSelector:@selector(compare:)];
+
+    // Try to generate a series of "update items" that safely transform
+    // the "old item list" into the "new item list".
```

Comment:
This is the core task.  The guiding principle is that `UICollectionView` incremental updates work best if you do deletes first, inserts second and updates last (because deletes and inserts affect the indices of subsequent operations).